include/diy/master.hpp

```diff
@@ -861,211 +870,307 @@ void
 
-      QueueRecord& in_qr  = current_incoming.map[to].records[from];
-      bool in_external  = block(lid(to)) == 0;
+  send_outgoing_queues(to_send, out_queues_limit, current_incoming, false);
 
-      if (in_external)
-          in_qr = it->second;
-      else
-      {
-          // load the queue
-          in_qr.size     = it->second.size;
-          in_qr.external = -1;
-
-          MemoryBuffer bb;
-          storage_->get(it->second.external, bb);
+  // kick requests
+  while(nudge());
 
-          current_incoming.map[to].queues[from].swap(bb);
-      }
-      ++current_incoming.received;
-    }
-    outgoing_[from].external_local.clear();
+  check_incoming_queues(false);
+}
 
-    if (outgoing_[from].external != -1)
-      load_outgoing(from);
-    to_send.pop_front();
+/* Remote communicator */
+
+// pseudocode for rexchange protocol based on NBX algorithm of Hoefler et al.,
+// Scalable Communication Protocols for Dynamic Sparse Data Exchange, 2010.
+//
+// rcomm_exchange()
+// {
+//      while (!done)
+//          while (sends_in_flight < limit_on_queues_in_memory and there are unprocessed queues)
+//              q = next outgoing queue (going over the in-memory queues first)
+//              if (q not in memory)
+//                  load q
+//              issend(q)
+//
+//           test all requests
+//           if (iprobe)
+//               recv
+//           if (barrier_active)
+//               if (test barrier)
+//                   done = true
+//           else
+//               if (all sends finished and all queues have been processed (including out-of-core queues))
+//                   ibarrier
+//                   barrier_active = true
+// }
+//
+void
+diy::Master::
+rcomm_exchange(ToSendList& to_send, int out_queues_limit)
+{
+    IncomingRound   &current_incoming   = incoming_[exchange_round_];
+    bool            done                = false;
+    bool            ibarr_act           = false;
+    mpi::request    ibarr_req;                      // mpi request associated with ibarrier
 
-    OutgoingQueues& outgoing = outgoing_[from].queues;
-    for (OutgoingQueues::iterator it = outgoing.begin(); it != outgoing.end(); ++it)
+    while (!done)
     {
-      BlockID to_proc = it->first;
-      int     to      = to_proc.gid;
-      int     proc    = to_proc.proc;
-
-      log->debug("Processing queue:      {} <- {} of size {}", to, from, outgoing_[from].queues[to_proc].size());
+        send_outgoing_queues(to_send, out_queues_limit, current_incoming, true);
 
-      // There may be local outgoing queues that remained in memory
-      if (proc == comm_.rank())     // sending to ourselves: simply swap buffers
-      {
-        log->debug("Moving queue in-place: {} <- {}", to, from);
+        // kick requests
+        while (nudge());                            // TODO: use nudge, or loop over test instead? (nudge removes items from inflight_sends)
```

Comment:
Removing items from `inflight_sends` is important for the `to_send.empty()` test below.